src/util.js

```diff
@@ -152,39 +152,49 @@ export function createRegExpRestore () {
     return function() {
         // Now we've snapshotted some properties, escape the lastMatch string
         let esc = /[.?*+^$[\]\\(){}|-]/g,
-            lm = regExpCache.lastMatch.replace(esc, '\\$&'),
-            reg = new List();
+            lastMatch = regExpCache.lastMatch.replace(esc, '\\$&'),
+            exprStr = '';
 
         // If any of the captured strings were non-empty, iterate over them all
         if (has) {
             for (let i = 1; i <= 9; i++) {
                 let m = regExpCache['$'+i];
 
                 // If it's empty, add an empty capturing group
-                if (!m)
-                    lm = '()' + lm;
-
+                if (!m) {
+                  exprStr += '(';
+                  lastMatch = ')' + lastMatch;
+                }
                 // Else find the string in lm and escape & wrap it to capture it
                 else {
                     m = m.replace(esc, '\\$&');
-                    lm = lm.replace(m, '(' + m + ')');
+                    exprStr += lastMatch.substring(0, lastMatch.indexOf(m)) + '(';
+                    lastMatch = m + ')' + lastMatch.substring(lastMatch.indexOf(m) + m.length);
                 }
-
-                // Push it to the reg and chop lm to make sure further groups come after
-                arrPush.call(reg, lm.slice(0, lm.indexOf('(') + 1));
-                lm = lm.slice(lm.indexOf('(') + 1);
             }
         }
 
-        let exprStr = arrJoin.call(reg, '') + lm;
+        exprStr += lastMatch;
 
         // Shorten the regex by replacing each part of the expression with a match
         // for a string of that exact length.  This is safe for the type of
         // expressions generated above, because the expression matches the whole
         // match string, so we know each group and each segment between capturing
         // groups can be matched by its length alone.
-        exprStr = exprStr.replace(/(\\\(|\\\)|[^()])+/g, (match) => {
```

Comment:
Second significant problem with the code.  This second operation on `exprStr` takes a series of characters other than unescaped parentheses and replaces it with `[\s\S]{length}`.  This allows our RegEx to work with much longer input strings.   However, the prior version of this regex does not handle *un*escaped parentheses that come immediately after escaped backslashes.  The new version is much more complicated, but accounts for sequences of escaped and unescaped backslashes and parentheses.

Looking at it now, I think the whole thing will be much more maintainable if we stop using a RegEx here and instead iterate over the characters in `exprStr` and check for escaped characters and match groups character-by-character.  Happy to work on that tomorrow under this PR or a separate one post-merge.